htlcswitch/switch.go

```diff
@@ -0,0 +1,599 @@
+package htlcswitch
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"crypto/sha256"
+
+	"github.com/go-errors/errors"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/routing"
+	"github.com/roasbeef/btcd/wire"
+	"github.com/roasbeef/btcutil"
+)
+
+var (
+	// ErrChannelLinkNotFound is used when channel link hasn't been found.
+	ErrChannelLinkNotFound = errors.New("channel link not found")
+)
+
+// ChannelCloseType is a enum which signals the type of channel closure the
+// peer should execute.
+type ChannelCloseType uint8
+
+const (
+	// CloseRegular indicates a regular cooperative channel closure
+	// should be attempted.
+	CloseRegular ChannelCloseType = iota
+
+	// CloseForce indicates that the channel should be forcefully closed.
+	// This entails the broadcast of the commitment transaction directly on
+	// chain unilaterally.
+	CloseForce
+
+	// CloseBreach indicates that a channel breach has been dtected, and
+	// the link should immediately be marked as unavailable.
+	CloseBreach
+)
+
+// ChanClose represents a request wto close a particular channel specified by
+// its outpoint.
+type ChanClose struct {
+	CloseType ChannelCloseType
+
+	ChanPoint *wire.OutPoint
+
+	Updates chan *lnrpc.CloseStatusUpdate
+	Err     chan error
+}
+
+// pendingPayment represents the payment which made by user and waits for
+// updates to be sent in channels about has it been rejected or processed
+// successfully.
+type pendingPayment struct {
+	err      chan error
+	preimage chan [sha256.Size]byte
+}
+
+// Switch is a central messaging bus for all incoming/outgoing htlc's.
+// The goal of the switch is forward the incoming/outgoing htlc messages from
+// one channel to another, and also propagate the settle/fail htlc messages
+// back to original requester. In order to better understand the whole view:
+//	alice			   bob				   carol
+//	server - <-connection-> - server - - <-connection-> - - - server
+//	 |		   	  |				    |
+//   alice htlc			bob htlc		       carol htlc
+//     switch			switch			         switch
+//	|			 |       \			   |
+//	|			 |        \			   |
+// alice       <-------->  first bob    second bob   <---------> carol
+// channel link	    	  channel link   channel link		channel link
+type Switch struct {
+	started  int32
+	shutdown int32
+	wg       sync.WaitGroup
+	quit     chan bool
+	mutex    sync.RWMutex
+
+	// pendingPayments is correspondence of payments and its hashes,
+	// which is used to save the payments which made by user.
+	pendingPayments map[[sha256.Size]byte]*pendingPayment
+
+	// circuits is an structure which is used to forward the settle/fail
+	// htlc updates back to the add htlc initiator.
+	circuits *circuitMap
+
+	// links is a map of channel output and channel link which manages
+	// this channel.
+	links map[wire.OutPoint]ChannelLink
+
+	// commands is used to propagate the commands to the handler goroutine.
+	commands chan interface{}
```

Comment:
Similar comment here to that left in the `link.go` file: the channel should be split up as it were before to restore the previous level of concurrency.